### ISSUE TITLE
valueGetter.assign is not a function

### ISSUE THREAD

**Issue by madebysoren** [vitalets/angular-xeditable#302]
Hi,

I get the following when saving changes in a table row.

```
TypeError: valueGetter.assign is not a function
    at EditableController.self.save (xeditable.js:791)
    at xeditable.js:1261
    at Object.r [as forEach] (angular.js:323)
    at Object.base.$save (xeditable.js:1260)
    at c (angular.js:1008)
    at applyCallback (xeditable.js:1539)
    at Object.then (xeditable.js:1534)
    at Object.checkSelf (xeditable.js:1249)
    at angular.js:1003
    at applyCallback (xeditable.js:1539)(anonymous function) @ angular.js:11655$get @ angular.js:8596$get.n.$apply @ angular.js:14567(anonymous function) @ xeditable.js:1453c @ angular.js:3032
```

Any idea of what can cause this?

Best,

Søren

**Comment by JessyH** [comment]
Hi, 

Same error here. I got this when i'm trying to include several variables in "editable-select".
Ex : 
"editable-select="table.number +' '+ table.SIMNickName +' '+ table.msisdn""

Hope it helps.

Jessy

**Comment by skukx** [comment]
I receive the same error when using a filter inside editable-text attr and try to save. so something like:

<code>
&lt;strong editable-text="name | uppercase"&gt;{{name | uppercase}}&lt;/strong&gt;
</code>

**Comment by jensenkd** [comment]
Any resolve on this?  I'm getting the same error.

**Comment by kalitas** [comment]
same here:
[Error] Error: undefined is not a function (evaluating 'valueGetter.assign($scope.$parent,
          self.useCopy ? angular.copy(self.scope.$data) : self.scope.$data)')
save@http://localhost:9000/bower_components/angular-xeditable/dist/js/xeditable.js:791:25
http://localhost:9000/bower_components/angular-xeditable/dist/js/xeditable.js:1261:22
forEach@http://localhost:9000/bower_components/angular/angular.js:336:24
$save@http://localhost:9000/bower_components/angular-xeditable/dist/js/xeditable.js:1260:22
http://localhost:9000/bower_components/angular/angular.js:1167:22
applyCallback@http://localhost:9000/bower_components/angular-xeditable/dist/js/xeditable.js:1539:19
then@http://localhost:9000/bower_components/angular-xeditable/dist/js/xeditable.js:1534:24
checkSelf@http://localhost:9000/bower_components/angular-xeditable/dist/js/xeditable.js:1249:16
http://localhost:9000/bower_components/angular/angular.js:1162:23
applyCallback@http://localhost:9000/bower_components/angular-xeditable/dist/js/xeditable.js:1539:19
then@http://localhost:9000/bower_components/angular-xeditable/dist/js/xeditable.js:1534:24
$submit@http://localhost:9000/bower_components/angular-xeditable/dist/js/xeditable.js:1238:14
http://localhost:9000/bower_components/angular-xeditable/dist/js/xeditable.js:1454:32
$eval@http://localhost:9000/bower_components/angular/angular.js:15846:28
$apply@http://localhost:9000/bower_components/angular/angular.js:15945:28
http://localhost:9000/bower_components/angular-xeditable/dist/js/xeditable.js:1453:29
dispatch@http://localhost:9000/bower_components/jquery/dist/jquery.js:4435:14
handle@http://localhost:9000/bower_components/jquery/dist/jquery.js:4121:33
	(anonymous function) (angular.js, line 12330)
	(anonymous function) (angular.js, line 9109)
	$apply (angular.js, line 15947)
	(anonymous function) (xeditable.js, line 1453)
	dispatch (jquery.js, line 4435)
	handle (jquery.js, line 4121)

**Comment by skukx** [comment]
This was happening for me when applying a filter.  I ended up using a workaround by applying the filter beforehand in the controller and then passing that variable into x-editable.
```javascript
app.controller('MyController', function() {
  vm = this;
  vm.originalValue = 'hello';
  vm.filteredValue = $filter('uppercase')(vm.originalValue);
}
```
```html
<a href="#" editable-text="vm.filteredValue">{{ vm.originalValue || 'empty' }}</a>
```

**Comment by nikki-forshaw** [comment]
Hi,

I'm getting the same issue, specifically on an angular xeditable date field. If I remove the buttons (buttons="no") then I don't get the issue, however, the date does not get saved.

Finding it very hard to find any info on this. Did anyone here find a solution aside from the $filter one (no filters in my code)?

Rather frustrated by it all!

TIA,
Nikki

**Comment by gfguthrie** [comment]
@nikki-forshaw I had this issue when the target of editable-text (or editable-select, etc.) was not a simple property. For example, I had `editable-select="item.arr.join(', ')"` and got the error, but when I changed it to `editable-select="item.arr"`, then the error went away.

The answer is then to use just simple properties that can be get & set, and do any transformation outside of the directive.

**Comment by sergio-ivanuzzo** [comment]
Maybe somebody will found this useful: I got this error when trying to submit form with boolean(true). I mean saving checkbox. So, for fixing this I set integer value instead of boolean.

**Comment by djandyr** [comment]
Same error, using angular filter on an ISO 8601 date on editable form, using editable-text. 

http://plnkr.co/edit/jvOzdvyHOQwz2fEtVBwA?p=preview

Is is possible for filters to be used? I assume the valueGetter parses the directive name, when a filter is used x-editable can not reference the editable element correctly.

**Comment by ckosloski** [comment]
@djandyr Did you try using a date picker instead of formatting a date for a text input?

**Comment by djandyr** [comment]
Thanks @ckosloski, using the date picker does [work](http://plnkr.co/edit/OzeTfddVZ4xDED33Ys7u?p=preview)

In general if I had a requirement for using a filter on a editable text element, is it supported?

**Comment by ckosloski** [comment]
Not currently.

**Comment by ChMrWang** [comment]
HI, I have a same issues, here is my code 
`$scope.saveNewGroupName = function () {
            if($scope.newGroupisEditing == false){ // It seems this function may be invoked even current group is not new group , it's strange but really happened
                return false;
            } else if(this.$data.length == 0){
                return "Please input a group name";
            } else {
                return $http.post("device/group/create", {
                    name: this.$data
                }).then(function (res) {
                    var device_group = res.data;
                    device_group.device = [];
                    $scope.deviceModel.deviceGroup.push(device_group);
                    $scope.newGroupisEditing = false;
                })
            }
        }`

**Comment by ckosloski** [comment]
Can you provide the code for your xeditable element?  If you are building the name of the model in the element, that currently is not supported.

**Comment by ChMrWang** [comment]
@ckosloski  Thanks for your quickly replication, I updated the xeditable element code in my first comments.

**Comment by ckosloski** [comment]
Why do you not have a model assigned to editable-text?

**Comment by ChMrWang** [comment]
@ckosloski  I will save the text to server and push the text to a array on saveNewGroupName() function, so a model assigned to editable-text is not needed for me. Is this the reason why this error appear?

**Comment by ckosloski** [comment]
Yes, you have to have a model.

**Comment by ChMrWang** [comment]
@ckosloski  OK, Thank you.

**Comment by jdmallen** [comment]
For others who might get this error, make sure that you're providing the model **without** the double curly braces "`{{}}`" in the `editable-select` argument. That was what caused it for me.

# Example
### This worked:
    <a href="" editable-select="myModel"
    	e-ng-options="o.id as o.name for o in myOptions"
    	onaftersave="updateModel()">{{myModel}}</a>

### This produced the error in this issue:
    <a href="" editable-select="{{myModel}}"
    	e-ng-options="o.id as o.name for o in myOptions"
    	onaftersave="updateModel()">{{myModel}}</a>

Also, if your options are an array of objects with subproperties like `id` and `name`, make sure the model you're updating is just a single property and not the whole object. That was the second problem I ran into. Notice that all the examples on [his docs](https://vitalets.github.io/angular-xeditable/#select-local) update just an `id` field on the `user` object-- `user` does not contain the whole of the option object.

In other words,

    $scope.user = {
    	status: 2
    };

instead of 

    $scope.user = {
    	status: {value: 2, text: 'status2'}
    };

Hope this helps.

**Comment by ckosloski** [comment]
Thanks for the comment @jdmallen 
@eugef I think you can close this issue.
